MH-12177 TimeZone non-threadsafe issue in scheduler-api, scheduler-impl

karendolan opened the pull request:
NOTE: this fix is to resolve the recurrence ical4j issues from the 4x bug-fixes merge back to develop/5x. This is *NOT* a fix for the 4x branch were, presumably, everything is working OK. (i.e. the bug fixes from 4x conflicted with new features added in develop)

This fixes the scheduler-impl unit test TimeZone issues with the Util's create periods method.
This includes a  first-Sunday-after-DST-time-change-within-a-period fix that is partially described in the link in the code comment. 
@gregorydlogan feel free to take and improve!

karendolan commented:
Hi @gregorydlogan This pull doesn't address the TimeZone setting activity in the index service unit tests. They don't fail for me and they don't impact a running Opencast. 

There is plenty to clean up on this pull. But, a big one is to consider moving the scheduler-impl UtilTest to the scheduler-api module where it belongs since the Util was moved there in an earlier pull.

karendolan commented:
The CI with the unit test shows the extra hour on fallback and  missing hour on spring forward SUNDAY in the DST time change, which is helpful. I'm going to add more junits  and attempt to finesse that SUNDAY fix .

@gregorydlogan  and @lkiesow , does this github CI test occur on a system east of the UTC?

```
[ERROR] calculateDSTSpringForwardChange(org.opencastproject.scheduler.impl.UtilTests)  Time elapsed: 0.184 s  <<< FAILURE!
java.lang.AssertionError: expected:<0> but was:<23>
	at org.opencastproject.scheduler.impl.UtilTests.doDSTChangeOverTest(UtilTests.java:163)
	at org.opencastproject.scheduler.impl.UtilTests.calculateDSTSpringForwardChange(UtilTests.java:113)

[ERROR] calculateDSTFallBackChange(org.opencastproject.scheduler.impl.UtilTests)  Time elapsed: 0.275 s  <<< FAILURE!
java.lang.AssertionError: expected:<0> but was:<1>
	at org.opencastproject.scheduler.impl.UtilTests.doDSTChangeOverTest(UtilTests.java:163)
	at org.opencastproject.scheduler.impl.UtilTests.calculateDSTFallBackChange(UtilTests.java:131)

```

gregorydlogan commented:
@karendolan Worthy of note: You reference the upstream github, but our version of ical4j dates back to the original funded Matterhorn project, and is actually a patched branch of 1.0.2 or something.  I'll file a PR against the DCE repo to fix that, among other things if I spot any.

karendolan commented:
@gregorydlogan wow, Ok. Also, in theory, there are a lot of new and useful Date and TimeZone tools built into Java 8.  I'm going to force push an iteration of this pull to show how crazy the work around is to get the TimeZone offsets correct across different DSTs and and ranges in the ical4j recurrence rules depending on the local TimeZone. It might be less crazy in a later version.

gregorydlogan commented:
Sounds good @karendolan.  I'll wait for that push and rebase my changes against it.

gregorydlogan commented on modules/scheduler-api/src/main/java/org/opencastproject/scheduler/api/Util.java:
```diff
@@ -44,58 +53,89 @@ private Util() {
-      } else { // Event doesn't start in DST
-        if (tz.inDaylightTime(d)) {
-          d.setTime(d.getTime() - tz.getDSTSavings()); // Adjust for Spring forward one hour
+    for (Object date : rRule.getRecur().getDates(periodStartTz, periodEndTz, net.fortuna.ical4j.model.parameter.Value.DATE_TIME)) {
```
Stepping through calculateDaysChange, I think there might be a logic bug in the RRule generation logic.  We have a rule, a start, and an end.  The start and end are exactly 5 days + duration apart, except that the timezone changes mid period, meaning the values are 5 days + duration - 1 hour apart.  In other words, periodStartTz is 0905, periodEndTz is 0*8*10, and the RRule says create a new event at 0900.  See the issue?  The rule is being followed to the letter, rather than what we meant.  So the question is, do we need two RRules?  Or should we do everything in UTC, and convert the RRule's time to UTC as well?

karendolan replied:
@gregorydlogan I think you are onto something, the latest mnode OSGI-compatible ical4j version of the fortuna ical4j seems to resolve some of the Sunday-schedule-in-the-middle-of-the-DST. But, it still has problems when the recurrence starts or ends in the middle of a DST change. The oddities are visible when you change your machine clock to  be on one side of the UTC and then test a recurrence with time zone on the other side of UTC and one of the scheduled dates falls on the DST day.

karendolan replied:
i.e. just when you think you've fixed the issue and then swap your machine time to be on the other side of the UTC, and the test schedule to be on the side, the issue re-manifests.

karendolan replied:
The ical4j code is on github. I suspect it is using machine time zone at some point in the recurrence calculation instead of sticking to UTC time.

gregorydlogan replied:
Looking at this more today, the recurrence logic internal to ical4j is totally broken.  For a period of 2016-03-25 - 2016-03-29 (the last item in calculateDaysChange in UtilTests), the recurrence code is generating five candidates:
- 2016-03-2**1**
- 2016-03-24
- 2016-03-25
- 2016-03-26
- 2016-03-27


Note the bolding on the first date.

I'll look further into their code tomorrow I guess.  I'm not sure if we're doing something wrong, but generating a date that far out of the possible range is very strange to me.  That's not an I-have-a-timezone bug...

gregorydlogan replied:
Further work, while my brains dribble out my ears.  The problematic code lives here:

```
getAbsWeekDays:1031, Recur (net.fortuna.ical4j.model)
getDayVariants:996, Recur (net.fortuna.ical4j.model)
getCandidates:818, Recur (net.fortuna.ical4j.model)
getDates:627, Recur (net.fortuna.ical4j.model)
getDates:520, Recur (net.fortuna.ical4j.model)
calculatePeriods:102, Util (org.opencastproject.scheduler.api)
generatePeriods:172, UtilTests (org.opencastproject.scheduler.api)
calculateDaysChange:90, UtilTests (org.opencastproject.scheduler.api)
```

Which lives [here](https://github.com/ical4j/ical4j/blob/develop/src/main/java/net/fortuna/ical4j/model/Recur.java#L1031).  I'm not 100% sure what exactly they're trying to do, but the issue stems from line 1033, where it sets the calendar to the first day of the start date's week.  This is 2018-03-21, which explains where *that* date is coming from.  It does not, however, explain how the rest of the dates are correct.  I'm too tired to continue tonight, but I'm very very sure that a large part of our current set of issues stem from the underlying ical4j library.

gregorydlogan replied:
This is now filed [here](https://github.com/ical4j/ical4j/issues/232) as well

karendolan replied:
Hi @gregorydlogan , using the latest ical4j and ensuring that the inbound recurrence dates are in the TZ that is being scheduled to (not UTC), the recurrence result dates and times are looking good and no fancy adjustments are needed anymore outside of the ical4j code. Test results are now identical when run on a system with local  TZ on either side of UTC. I'm doing some more sanity checks before pushing to the t/MH-12177-TimeZone-Issue-scheduler-dev branch. I've locally rebased that branch to OC develop. But this commit should be applicable to 4x, 5x.

smarquard commented:
Can this be targeted at 5.x rather than develop? It is an ongoing issue that affects production systems, and Opencast's scheduling and conflict detection is not trustworthy without this fix. 6.x is too late.

gregorydlogan commented:
@smarquard IIRC TimeZone.setDefault code actually came from 4.x, so I'm hoping to target it there and merge it forward.

smarquard commented:
The setDefault code has been in use since 2.x, but it's probably not feasible to target at earlier than 4.x.

karendolan commented on modules/scheduler-api/src/main/java/org/opencastproject/scheduler/api/Util.java:
```diff
@@ -21,81 +21,117 @@
    *
-   *  NOTE: Do not modify this without making the same modifications to the copy of this method in IndexServiceImplTest
-   *  I would have moved this to the scheduler-api bundle, but that would introduce a circular dependency :(
+   *TODO
```
The NOTE can be removed. If forgot to do this TODO.

Commit:
MH-12177 fix mvn site

karendolan commented on modules/scheduler-impl/src/main/java/org/opencastproject/scheduler/endpoint/SchedulerRestService.java:
```diff
@@ -1081,7 +1082,7 @@ public Response getEventsAsXml(@QueryParam("agent") String device, @QueryParam("
           @RestParameter(name = "rrule", description = "Rule for recurrent conflicting, specified as: \"FREQ=WEEKLY;BYDAY=day(s);BYHOUR=hour;BYMINUTE=minute\". FREQ is required. BYDAY may include one or more (separated by commas) of the following: SU,MO,TU,WE,TH,FR,SA.", isRequired = false, type = Type.STRING),
           @RestParameter(name = "duration", description = "If recurrence rule is specified duration of each conflicting period, in milliseconds", isRequired = false, type = Type.INTEGER),
-          @RestParameter(name = "timezone", description = "The timezone of the capture device", isRequired = false, type = Type.STRING) }, reponses = {
+          @RestParameter(name = "timezone", description = "The timezone of the capture device", isRequired = true, type = Type.STRING) }, reponses = {
```
I missed that if no rule is passed, the timezone doesn't need to be passed. I'll change the required back to false.

karendolan commented:
@gregorydlogan I did not yet have a chance to test the pull with Date assumptions made in the admin-UI. I suspect, but have not verified, that contains assumptions of dates coming in as UTC vs the time designated by the TZ param (if one is passed) or local TZ (if a TZ param is not passed). When the dates eventually go into a recurrence, they need to be correctly adjusted and associated to the TZ in order for the ical4j algorithms to work. 

It may be possible to do more work on this pull to allow inbound UTC dates and then convert them to the TZ of choice before performing ical4j calculations. I tried this and it got complex quickly. For example, if the inbound "day" to schedule is MON, and the code takes the hour-of-day to start the event at 11pm, and the TZ is UTC-3, every time that Date is assessed for hour-of-day or day-of-week, or day-of-month, it needs extra care to ensure it's converted into TZ first. The TZ day-of-week, day-of-month is one unit behind the the UTC day-of-week, day-of-month. The ical4j, joda time, java.time Date and DateTime classes are not as compatible as one would think. Using "Calendar" and "GeorgianCalendar" to wrap Date in a TZ context helps to make schedule related work less hazardous.

KatrinIhler commented:
@karendolan @gregorydlogan What is supposed to happen here? Is this on hold currently?

gregorydlogan commented:
@KatrinIhler little to no progress.  The underlying library has an issue, and it's quite a complex piece of code.  Since no one else is complaining here I'd say this is on hold until I can find a resolution.

karendolan commented:
@gregorydlogan Hi, meanwhile I'll rebase this onto latest develop.

Commit:
MH-12177 TimeZone non-threadsafe issue in scheduler-api, scheduler-impl
         update ical4j lib (from net.fortuna.ical4j 1.0-rc2-patched to org.mnode.ical4j 2.2.0)
         Recurrence inbound start and end dates must be in schedule to TimeZone
            (not converted UTC) for correct days of week and DST changeover to work.
         moved scheduler Util test to scheduler-api (current home of Scheduler Util)

Commit:
MH-12177 schedule to specific timezone or default to local timezone

karendolan commented:
Hi, based on the recent dev list "[OC Dev] Tips for debug build errors" thread and MH-13264. I'm doing a quick revisit of the goal behind this pull using the latest r/6.x, related libs, and current Opencast internal & external timezone/UTC viewpoints (based on list emails and 6x pulls). 

To recap: The core problem is that bulk scheduling over daylight savings change-overs choke depending on which side of UTC your Opencst system lives in.

gregorydlogan commented:
Hi Karen

Can you reach out to the folks in that thread and coordinate? It sounds
like Duncan is working on something that will either conflict or be
helpful.

G

On Tue, Dec 4, 2018, 08:15 Karen Dolan <notifications@github.com wrote:

> Hi, based on the recent dev list "[OC Dev] Tips for debug build errors"
> thread and MH-13264. I'm doing a quick revisit of the goal behind this pull
> using the latest r/6.x, related libs, and current Opencast internal &
> external timezone/UTC viewpoints (based on list emails and 6x pulls).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/opencast/opencast/pull/189#issuecomment-444114259>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFYNiYAWx7RVgCuS3RQ0jaLg3STBQmdPks5u1oOLgaJpZM4TSpbq>
> .
>

karendolan commented:
@gregorydlogan @KatrinIhler @smarquard I applied the current state of this pull to r/6.x https://github.com/opencast/opencast/pull/604. I recommend closing out this pull in favor of that one to start clean with the comments since so much had changed over the course of this pull.

karendolan commented:
Closing this pull request in favor of updated pull at https://github.com/opencast/opencast/pull/604
I am unable to reproduce the UTC/non-DST timeZone issues that @gregorydlogan discovered in the underlying ical4j library using pull #604. Possibly, because a TZ object is always passed to the recurrence rule via #604 and is not being passed in the existing code.
